### PR TITLE
line chart: support relative time

### PR DESCRIPTION
**Pull request by stephanwlee**
Relative time is where we lay data points out relative to the wallTime
of the first datum. This change adds the necessary data transformation
in the scalar_card and add its support in the tooltip.

Future PR todo:
- pipe the custom formatter to tooltip + axis renderer.

### CODE REVIEW

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_container.ts [hunk 1]
```diff
@@ -283,21 +300,28 @@ export class ScalarCardContainer implements CardRenderer, OnInit {
         this.store.select(getMetricsScalarSmoothing),
         this.store.select(getMetricsXAxisType)
       ),
-      switchMap(([runsData, smoothing, xType]) => {
+      switchMap(([runsData, smoothing, xScaleType]) => {
```
Comment (bmd3k) [inline tensorboard/webapp/metrics/views/card_renderer/scalar_card_container.ts:303]: xScaleType => xAxisType?

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_container.ts [hunk 2]
```diff
@@ -283,21 +300,28 @@ export class ScalarCardContainer implements CardRenderer, OnInit {
...
-              return {
+          points = points.map((point) => {
+            // Convert wallTime in seconds to milliseconds.
+            // TODO(stephanwlee): when the legacy line chart is removed, do the conversion at
+            // the effects.
+            const wallTime = point.wallTime * 1000;
+            const x = xScaleType === XAxisType.STEP ? point.x : wallTime;
+            return {...point, x, wallTime};
+          });
+          switch (xScaleType) {
```
Comment (bmd3k) [inline tensorboard/webapp/metrics/views/card_renderer/scalar_card_container.ts:313]: An if statement would be clearer:

if (xScaleType === XAxisType.RELATIVE and points.length) {
  // do stuff.
}

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_container.ts [hunk 3]
```diff
@@ -180,6 +182,21 @@ export class ScalarCardContainer implements CardRenderer, OnInit {
   readonly tooltipSort$ = this.store.select(getMetricsTooltipSort);
   readonly ignoreOutliers$ = this.store.select(getMetricsIgnoreOutliers);
   readonly xAxisType$ = this.store.select(getMetricsXAxisType);
+  readonly newXAxisType$ = this.xAxisType$.pipe(
```
Comment (bmd3k) [inline tensorboard/webapp/metrics/views/card_renderer/scalar_card_container.ts:185]: newXAxisType => xScaleType?

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_component.ts [hunk 4]
```diff
@@ -122,6 +123,8 @@ export class ScalarCardComponent {
   @Input() showFullSize!: boolean;
   @Input() isPinned!: boolean;
 
+  @Input() newXAxisType!: ScaleType;
```
Comment (bmd3k) [inline tensorboard/webapp/metrics/views/card_renderer/scalar_card_component.ts:126]: newXAxisType => xScaleType?  (XAxisType seems to be something different than ScaleType in your data model. I don't really understand what the "new" is supposed to mean but I guess it means "ScaleType is the new XAxisType")

File: tensorboard/webapp/widgets/line_chart_v2/line_chart_component.ts [hunk 5]
```diff
@@ -97,6 +98,12 @@ export class LineChartComponent implements AfterViewInit, OnChanges, OnDestroy {
   @Input()
   yScaleType: ScaleType = ScaleType.LINEAR;
 
+  @Input()
+  customXFormatter?: Formatter;
```
Comment (bmd3k) [inline tensorboard/webapp/widgets/line_chart_v2/line_chart_component.ts:102]: I don't see these used. Or is this what the TODO refers to? ("pipe the custom formatter to tooltip + axis renderer.").  It's a bit difficult to judge, as reviewer, whether this makes much sense - maybe consider deferring changes to line_chart_component to the future PR?

**Review by bmd3k — APPROVED** [review]

File: tensorboard/webapp/widgets/line_chart_v2/line_chart_component.ts (hunk 5 above)
Comment (stephanwlee) [reply tensorboard/webapp/widgets/line_chart_v2/line_chart_component.ts:102]: Thanks for noting that. I forgot to write the glue code. Done.

**Pull request merged**